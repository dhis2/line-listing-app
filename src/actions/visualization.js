import { getUiDimensionType } from '../modules/dimensionConstants.js'
import { formatDimensionId } from '../modules/dimensionId.js'
import {
    getDynamicTimeDimensionsMetadata,
    getProgramAsMetadata,
} from '../modules/metadata.js'
import { getDimensionMetadataFromVisualization } from '../modules/visualization.js'
import {
    SET_VISUALIZATION,
    CLEAR_VISUALIZATION,
} from '../reducers/visualization.js'

export const acSetVisualization = (value) => {
    const { outputType, program, programStage } = value
    const timeDimensions = getDynamicTimeDimensionsMetadata(
        program,
        programStage
    )
    const collectedMetadata = getDimensionMetadataFromVisualization(value) || {}
    const dimensions = [
        ...(value.columns || []),
        ...(value.rows || []),
        ...(value.filters || []),
    ]

    const metadata = dimensions.reduce((md, dimension) => {
        if (!collectedMetadata[dimension?.dimension]) {
            return md
        }

        const prefixedId = formatDimensionId({
            dimensionId: dimension.dimension,
            programStageId: dimension.programStage?.id,
            programId: dimension.program?.id,
            outputType,
        })

        md.push({
            [prefixedId]: {
                id: prefixedId,
                name: collectedMetadata[dimension.dimension].name,
                valueType: dimension.valueType,
                optionSet: dimension.optionSet?.id,
                dimensionType: getUiDimensionType({
                    dimensionId: prefixedId,
                    dimensionType: dimension.dimensionType,
                }),
            },
        })

        return md
    }, [])

    // program
    if (program) {
        metadata.push(getProgramAsMetadata(program))
    }

    // program stage
    if (programStage) {
        metadata.push({
            [programStage.id]: programStage,
        })
    }

    // time dimensions
    Object.keys(timeDimensions).forEach((timeDimensionId) => {
        metadata.push({
            [timeDimensionId]: timeDimensions[timeDimensionId],
        })
    })

    return {
        type: SET_VISUALIZATION,
        value,
        metadata,
    }
}

export const acClearVisualization = () => ({
    type: CLEAR_VISUALIZATION,
})
