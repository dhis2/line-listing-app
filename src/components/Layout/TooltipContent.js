import { ouIdHelper } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { getConditions } from '../../modules/conditions.js'
import {
    DIMENSION_TYPE_CATEGORY,
    DIMENSION_TYPE_CATEGORY_OPTION_GROUP_SET,
    DIMENSION_TYPE_ORGANISATION_UNIT_GROUP_SET,
    DIMENSION_TYPE_STATUS,
    DIMENSION_TYPE_PERIOD,
    DIMENSION_TYPE_OU,
} from '../../modules/dimensionConstants.js'
import { sGetMetadata } from '../../reducers/metadata.js'
import {
    sGetUiConditionsByDimension,
    sGetUiItemsByDimension,
} from '../../reducers/ui.js'
import styles from './styles/Tooltip.module.css'

const labels = {
    noneSelected: () => i18n.t('None selected'),
    allSelected: () => i18n.t('Showing all values for this dimension'),
    oneOverLimit: () => i18n.t('And 1 other...'),
    nOverLimit: (n) =>
        i18n.t('And {{numberOfItems}} others...', {
            numberOfItems: n,
        }),
}

const renderLimit = 5

export const TooltipContent = ({
    dimension,
    itemIds,
    metadata,
    conditions,
}) => {
    const getNameList = (idList, label, metadata) =>
        idList.reduce(
            (levelString, levelId, index) =>
                `${levelString}${index > 0 ? `, ` : ``}${
                    metadata[levelId] ? metadata[levelId].name : levelId
                }`,
            `${label}: `
        )

    const getItemDisplayNames = () => {
        const levelIds = []
        const groupIds = []
        const itemDisplayNames = []

        itemIds.forEach((id) => {
            if (ouIdHelper.hasLevelPrefix(id)) {
                levelIds.push(ouIdHelper.removePrefix(id))
            } else if (ouIdHelper.hasGroupPrefix(id)) {
                groupIds.push(ouIdHelper.removePrefix(id))
            } else {
                itemDisplayNames.push(metadata[id] ? metadata[id].name : id)
            }
        })

        levelIds.length &&
            itemDisplayNames.push(
                getNameList(levelIds, i18n.t('Levels'), metadata)
            )

        groupIds.length &&
            itemDisplayNames.push(
                getNameList(groupIds, i18n.t('Groups'), metadata)
            )

        return itemDisplayNames
    }

    const renderItems = (itemDisplayNames) => {
        const itemsToRender = itemDisplayNames
            .slice(0, renderLimit)
            .map((name) => (
                <li key={`${dimension.id}-${name}`} className={styles.item}>
                    {name}
                </li>
            ))

        if (itemDisplayNames.length > renderLimit) {
            itemsToRender.push(
                <li
                    key={`${dimension.id}-render-limit`}
                    className={styles.item}
                >
                    {itemDisplayNames.length - renderLimit === 1
                        ? labels.oneOverLimit
                        : labels.nOverLimit(
                              itemDisplayNames.length - renderLimit
                          )}
                </li>
            )
        }

        return itemsToRender
    }

    const renderConditions = () =>
        renderItems(getConditions({ conditions, metadata, dimension }))

    const renderNoItemsLabel = () => (
        <li
            key={`${dimension.id}-${labels.noneSelected()}`}
            className={styles.item}
        >
            {labels.noneSelected()}
        </li>
    )

    const renderAllItemsLabel = () => (
        <li
            key={`${dimension.id}-${labels.allSelected()}`}
            className={styles.item}
        >
            {labels.allSelected()}
        </li>
    )

    const itemDisplayNames = Boolean(itemIds.length) && getItemDisplayNames()

    switch (dimension.dimensionType) {
        case DIMENSION_TYPE_CATEGORY:
        case DIMENSION_TYPE_CATEGORY_OPTION_GROUP_SET:
        case DIMENSION_TYPE_ORGANISATION_UNIT_GROUP_SET:
        case DIMENSION_TYPE_STATUS:
            return (
                <ul className={styles.list}>
                    {itemDisplayNames
                        ? renderItems(itemDisplayNames)
                        : renderAllItemsLabel()}
                </ul>
            )

        case DIMENSION_TYPE_PERIOD:
        case DIMENSION_TYPE_OU:
            return (
                <ul className={styles.list}>
                    {itemDisplayNames
                        ? renderItems(itemDisplayNames)
                        : renderNoItemsLabel()}
                </ul>
            )

        default: {
            return (
                <ul className={styles.list}>
                    {conditions?.condition || conditions?.legendSet
                        ? renderConditions(conditions)
                        : renderAllItemsLabel()}
                </ul>
            )
        }
    }
}

TooltipContent.propTypes = {
    conditions: PropTypes.object.isRequired,
    dimension: PropTypes.object.isRequired,
    metadata: PropTypes.object.isRequired,
    itemIds: PropTypes.array,
}

const mapStateToProps = (state, ownProps) => ({
    metadata: sGetMetadata(state),
    itemIds: sGetUiItemsByDimension(state, ownProps.dimension.id) || [],
    conditions: sGetUiConditionsByDimension(state, ownProps.dimension.id) || {},
})

export default connect(mapStateToProps)(TooltipContent)
