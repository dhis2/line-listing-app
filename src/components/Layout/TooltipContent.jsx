import {
    DIMENSION_TYPE_CATEGORY,
    DIMENSION_TYPE_CATEGORY_OPTION_GROUP_SET,
    DIMENSION_TYPE_DATA_ELEMENT,
    ouIdHelper,
    DIMENSION_TYPE_ORGANISATION_UNIT,
    DIMENSION_TYPE_PERIOD,
    DIMENSION_TYPE_ORGANISATION_UNIT_GROUP_SET,
    AXIS_ID_FILTERS,
} from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'
import React from 'react'
import { useSelector } from 'react-redux'
import {
    isStartEndDate,
    useLocalizedStartEndDateFormatter,
} from '../../modules/dates.js'
import { DIMENSION_TYPE_STATUS } from '../../modules/dimensionConstants.js'
import { extractDimensionIdParts } from '../../modules/dimensionId.js'
import { sGetMetadata } from '../../reducers/metadata.js'
import { sGetUiInputType, sGetUiItemsByDimension } from '../../reducers/ui.js'
import styles from './styles/Tooltip.module.css'

const renderLimit = 5

export const TooltipContent = ({ dimension, conditionsTexts, axisId }) => {
    const metadata = useSelector(sGetMetadata)
    const inputType = useSelector(sGetUiInputType)
    const itemIds = useSelector((state) =>
        sGetUiItemsByDimension(state, dimension.id)
    )
    const formatStartEndDate = useLocalizedStartEndDateFormatter()

    const { programStageId, programId } = extractDimensionIdParts(
        dimension.id,
        inputType
    )

    const stageName = dimension.stageName || metadata[programStageId]?.name
    const programName = metadata[programId]?.name

    const getNameList = (idList, label, metadata) =>
        idList.reduce(
            (levelString, levelId, index) =>
                `${levelString}${index > 0 ? `, ` : ``}${
                    metadata[levelId] ? metadata[levelId].name : levelId
                }`,
            `${label}: `
        )

    const getItemDisplayNames = () => {
        const levelIds = []
        const groupIds = []
        const itemDisplayNames = []

        itemIds.forEach((id) => {
            if (ouIdHelper.hasLevelPrefix(id)) {
                levelIds.push(ouIdHelper.removePrefix(id))
            } else if (ouIdHelper.hasGroupPrefix(id)) {
                groupIds.push(ouIdHelper.removePrefix(id))
            } else {
                const { dimensionId } = extractDimensionIdParts(id, inputType)
                itemDisplayNames.push(
                    isStartEndDate(dimensionId)
                        ? formatStartEndDate(dimensionId)
                        : metadata[dimensionId]?.name ?? id
                )
            }
        })

        levelIds.length &&
            itemDisplayNames.push(
                getNameList(levelIds, i18n.t('Levels'), metadata)
            )

        groupIds.length &&
            itemDisplayNames.push(
                getNameList(groupIds, i18n.t('Groups'), metadata)
            )

        return itemDisplayNames
    }

    const renderItems = (itemDisplayNames = []) => {
        if (itemDisplayNames.some((name) => !name)) {
            return null
        }
        const itemsToRender = itemDisplayNames
            .slice(0, renderLimit)
            .map((name) => (
                <li key={`${dimension.id}-${name}`} className={styles.item}>
                    {name}
                </li>
            ))

        const numberOverRenderLimit = itemDisplayNames.length - renderLimit
        if (numberOverRenderLimit > 0) {
            itemsToRender.push(
                <li
                    key={`${dimension.id}-render-limit`}
                    className={styles.item}
                >
                    {i18n.t('And {{count}} other...', {
                        count: numberOverRenderLimit,
                        defaultValue: 'And {{count}} other...',
                        defaultValue_plural: 'And {{count}} others...',
                    })}
                </li>
            )
        }

        return itemsToRender
    }

    const renderNoItemsLabel = () => (
        <li key={`${dimension.id}-none-selected`} className={styles.item}>
            {i18n.t('None selected')}
        </li>
    )

    const renderStageName = () =>
        stageName && (
            <li className={styles.item}>
                {i18n.t('Program stage: {{- stageName}}', {
                    stageName,
                    nsSeparator: '^^',
                })}
            </li>
        )

    const renderProgramName = () =>
        programName && (
            <li className={styles.item}>
                {i18n.t('Program: {{- programName}}', {
                    programName,
                    nsSeparator: '^^',
                })}
            </li>
        )

    const renderItemsSection = (itemsList) => {
        if (itemsList.length) {
            return renderItems(itemsList)
        } else {
            if (axisId === AXIS_ID_FILTERS) {
                return renderNoItemsLabel()
            } else {
                return (
                    <li
                        key={`${dimension.id}-all-selected`}
                        className={styles.item}
                    >
                        {i18n.t('Showing all values for this dimension')}
                    </li>
                )
            }
        }
    }

    const itemDisplayNames = Boolean(itemIds.length) && getItemDisplayNames()

    switch (dimension.dimensionType) {
        case DIMENSION_TYPE_CATEGORY:
        case DIMENSION_TYPE_CATEGORY_OPTION_GROUP_SET:
        case DIMENSION_TYPE_ORGANISATION_UNIT_GROUP_SET:
        case DIMENSION_TYPE_STATUS:
            return (
                <ul className={styles.list} data-test="tooltip-content">
                    {renderProgramName()}
                    {renderItemsSection(itemDisplayNames)}
                </ul>
            )
        case DIMENSION_TYPE_PERIOD:
        case DIMENSION_TYPE_ORGANISATION_UNIT:
            return (
                <ul className={styles.list} data-test="tooltip-content">
                    {renderProgramName()}
                    {itemDisplayNames
                        ? renderItems(itemDisplayNames)
                        : renderNoItemsLabel()}
                </ul>
            )
        case DIMENSION_TYPE_DATA_ELEMENT: {
            return (
                <ul className={styles.list} data-test="tooltip-content">
                    {renderProgramName()}
                    {renderStageName()}
                    {renderItemsSection(conditionsTexts)}
                </ul>
            )
        }
        default: {
            return (
                <ul className={styles.list} data-test="tooltip-content">
                    {renderProgramName()}
                    {renderItemsSection(conditionsTexts)}
                </ul>
            )
        }
    }
}

TooltipContent.propTypes = {
    conditionsTexts: PropTypes.array.isRequired,
    dimension: PropTypes.object.isRequired,
    axisId: PropTypes.string,
}

export default TooltipContent
