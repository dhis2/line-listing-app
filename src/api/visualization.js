import { getDimensionMetadataFields } from '../modules/visualization.js'

const dimensionFields = () =>
    'dimension,dimensionType,filter,program[id],programStage[id],optionSet[id],valueType,legendSet[id],repetition,items[dimensionItem~rename(id)]'

export const VISUALIZATION_QUERY = {
    eventVisualization: {
        resource: 'eventVisualizations',
        id: ({ id }) => id,
        // TODO: check if this list is what we need/want (copied from old ER)
        params: ({ nameProp }) => ({
            fields: [
                '*',
                `columns[${dimensionFields}]`,
                `rows[${dimensionFields}]`,
                `filters[${dimensionFields}]`,
                `program[id,programType,${nameProp}~rename(name),displayEnrollmentDateLabel,displayIncidentDateLabel,displayIncidentDate,programStages[id,displayName~rename(name),repeatable]]`,
                'programStage[id,displayName~rename(name),displayExecutionDateLabel,displayDueDateLabel,hideDueDate,repeatable]',
                `programDimensions[id,${nameProp}~rename(name),enrollmentDateLabel,incidentDateLabel,programType,displayIncidentDate,displayEnrollmentDateLabel,displayIncidentDateLabel,programStages[id,${nameProp}~rename(name),repeatable,hideDueDate,displayExecutionDateLabel,displayDueDateLabel]]`,
                'access',
                'href',
                ...getDimensionMetadataFields(),
                'dataElementDimensions[legendSet[id,name],dataElement[id,name]]',
                'legend[set[id,displayName],strategy,style,showKey]',
                'trackedEntityType[id,displayName~rename(name)]',
                '!interpretations',
                '!userGroupAccesses',
                '!publicAccess',
                '!displayDescription',
                '!rewindRelativePeriods',
                '!userOrganisationUnit',
                '!userOrganisationUnitChildren',
                '!userOrganisationUnitGrandChildren',
                '!externalAccess',
                '!relativePeriods',
                '!columnDimensions',
                '!rowDimensions',
                '!filterDimensions',
                '!organisationUnitGroups',
                '!itemOrganisationUnitGroups',
                '!indicators',
                '!dataElements',
                '!dataElementOperands',
                '!dataElementGroups',
                '!dataSets',
                '!periods',
                '!organisationUnitLevels',
                '!organisationUnits',
                '!user',
            ],
        }),
    },
}

const VISUALIZATION_NAME_DESC_QUERY = {
    eventVisNameDesc: {
        resource: 'eventVisualizations',
        id: ({ id }) => id,
        params: {
            fields: 'name,displayName,description,displayDescription',
        },
    },
}

const VISUALIZATION_SUBSCRIBERS_QUERY = {
    eventVisualizationSubscribers: {
        resource: 'eventVisualizations',
        id: ({ id }) => id,
        params: {
            fields: ['subscribers'],
        },
    },
}

export const apiFetchVisualization = ({ engine, id, nameProp }) =>
    engine.query(VISUALIZATION_QUERY, {
        variables: { id, nameProp },
    })

export const apiFetchVisualizationNameDesc = ({ engine, id }) =>
    engine
        .query(VISUALIZATION_NAME_DESC_QUERY, {
            variables: { id },
        })
        .then(({ eventVisNameDesc }) => eventVisNameDesc)

export const apiFetchVisualizationSubscribers = ({ engine, id }) =>
    engine
        .query(VISUALIZATION_SUBSCRIBERS_QUERY, {
            variables: { id },
        })
        .then(
            ({ eventVisualizationSubscribers }) => eventVisualizationSubscribers
        )
