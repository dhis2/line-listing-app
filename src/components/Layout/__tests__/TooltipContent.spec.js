import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { OUTPUT_TYPE_EVENT } from '../../../modules/visualization.js'
import { TooltipContent } from '../TooltipContent.js'

const mockStore = configureMockStore()

const inputTypeEvent = {
    input: {
        type: OUTPUT_TYPE_EVENT,
    },
}

describe('TooltipContent', () => {
    test('OU: 1 id and 1 level', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {
                    ou: ['theOuId', 'LEVEL-level1Id'],
                },
            },
            metadata: {
                theOuId: { name: 'My OU' },
                level1Id: { name: 'My level 1' },
            },
        }

        const dimension = {
            id: 'ou',
            dimensionType: 'ORGANISATION_UNIT',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('OU: 1 id and 1 level, and missing metadata', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {
                    ou: ['theOuId', 'LEVEL-level1Id'],
                },
            },
            metadata: {},
        }

        const dimension = {
            id: 'ou',
            dimensionType: 'ORGANISATION_UNIT',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('OU: 1 id and 3 groups', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {
                    ou: [
                        'theOuId',
                        'OU_GROUP-group1Id',
                        'OU_GROUP-group2Id',
                        'OU_GROUP-group3Id',
                    ],
                },
            },
            metadata: {
                theOuId: { name: 'OU Equestria' },
                group1Id: { name: 'All the blue ponies' },
                group2Id: { name: 'All the yellow ponies' },
                group3Id: { name: 'All the purple ponies' },
            },
        }

        const dimension = {
            id: 'ou',
            dimensionType: 'ORGANISATION_UNIT',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('OU: none selected', () => {
        const store = {
            ui: { ...inputTypeEvent, itemsByDimension: {} },
            metadata: {},
        }

        const dimension = {
            id: 'ou',
            dimensionType: 'ORGANISATION_UNIT',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Period: none selected', () => {
        const store = {
            ui: { ...inputTypeEvent, itemsByDimension: {} },
            metadata: {
                eventDate: { name: 'Report compilation date' },
            },
        }

        const dimension = {
            id: 'eventDate',
            dimensionType: 'PERIOD',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Period: 2 selected', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {
                    eventDate: ['202205', '202207'],
                },
            },
            metadata: {
                202205: {
                    id: '202205',
                    name: 'May 2022',
                },
                202207: {
                    id: '202207',
                    name: 'July 2022',
                },
            },
        }

        const dimension = {
            id: 'eventDate',
            dimensionType: 'PERIOD',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('COGS: 2 items', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {
                    COGS_id: ['cogs1Id', 'cogs2Id'],
                },
            },
            metadata: {
                cogs1Id: {
                    id: 'cogs1Id',
                    name: 'COGS item 1',
                    code: 'cogs1Code',
                },
                cogs2Id: {
                    id: 'cogs2Id',
                    name: 'COGS item 2',
                    code: 'cogs2Code',
                },
            },
        }

        const dimension = {
            id: 'COGS_id',
            dimensionType: 'CATEGORY_OPTION_GROUP_SET',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: optionSet, 2 conditions but name missing for one', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {},
            },
            metadata: {},
        }

        const dimension = {
            id: 'myDataElementId',
            name: 'My fascinating Data Element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'TEXT',
            optionSet: 'optionSetId',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={true}
                    conditionsTexts={[undefined, 'Condition 2']}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: with stage and 1 condition', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {},
            },
            metadata: {
                stageId: { name: 'The first stage' },
                'stageId.dataElementId': { name: 'Lucky numbers' },
            },
        }

        const dimension = {
            id: 'stageId.dataElementId',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
            name: 'Lucky numbers',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={true}
                    conditionsTexts={['Condition 1']}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: no conditions', () => {
        const store = {
            ui: { ...inputTypeEvent, itemsByDimension: {} },
            metadata: {},
        }

        const dimension = {
            id: 'dataElementId',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={false}
                    conditionsTexts={[]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: 2 conditions', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {},
            },
            metadata: {},
        }

        const dimension = {
            id: 'dimensionId',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={true}
                    conditionsTexts={['Condition 1', 'Condition 2']}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: 6 conditions', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {},
            },
            metadata: {},
        }

        const dimension = {
            id: 'dimensionId',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={true}
                    conditionsTexts={[
                        'Condition 1',
                        'Condition 2',
                        'Condition 3',
                        'Condition 4',
                        'Condition 5',
                        'Condition 6',
                    ]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
    // TODO: figure out why plural translations are not being returned from i18n.t() in Jest tests
    // The snapshot should say "And 3 others..."
    test('Data element: 8 conditions', () => {
        const store = {
            ui: {
                ...inputTypeEvent,
                itemsByDimension: {},
            },
            metadata: {},
        }

        const dimension = {
            id: 'dimensionId',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <TooltipContent
                    dimension={dimension}
                    hasConditions={true}
                    conditionsTexts={[
                        'Condition 1',
                        'Condition 2',
                        'Condition 3',
                        'Condition 4',
                        'Condition 5',
                        'Condition 6',
                        'Condition 7',
                        'Condition 8',
                    ]}
                />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })
})
