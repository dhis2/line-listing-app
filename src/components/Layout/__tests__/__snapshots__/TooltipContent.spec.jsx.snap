// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipContent COGS: 2 items 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      COGS item 1
    </li>
    <li
      class="item"
    >
      COGS item 2
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Data element: 2 conditions 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      Condition 1
    </li>
    <li
      class="item"
    >
      Condition 2
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Data element: 6 conditions 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      Condition 1
    </li>
    <li
      class="item"
    >
      Condition 2
    </li>
    <li
      class="item"
    >
      Condition 3
    </li>
    <li
      class="item"
    >
      Condition 4
    </li>
    <li
      class="item"
    >
      Condition 5
    </li>
    <li
      class="item"
    >
      And 1 other...
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Data element: 8 conditions 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      Condition 1
    </li>
    <li
      class="item"
    >
      Condition 2
    </li>
    <li
      class="item"
    >
      Condition 3
    </li>
    <li
      class="item"
    >
      Condition 4
    </li>
    <li
      class="item"
    >
      Condition 5
    </li>
    <li
      class="item"
    >
      And 3 other...
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Data element: no conditions 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      Showing all values for this dimension
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Data element: optionSet, 2 conditions but name missing for one 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  />
</div>
`;

exports[`TooltipContent Data element: with stage and 1 condition 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      Program stage: The first stage
    </li>
    <li
      class="item"
    >
      Condition 1
    </li>
  </ul>
</div>
`;

exports[`TooltipContent OU: 1 id and 1 level 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      My OU
    </li>
    <li
      class="item"
    >
      Levels: My level 1
    </li>
  </ul>
</div>
`;

exports[`TooltipContent OU: 1 id and 1 level, and missing metadata 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      theOuId
    </li>
    <li
      class="item"
    >
      Levels: level1Id
    </li>
  </ul>
</div>
`;

exports[`TooltipContent OU: 1 id and 3 groups 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      OU Equestria
    </li>
    <li
      class="item"
    >
      Groups: All the blue ponies, All the yellow ponies, All the purple ponies
    </li>
  </ul>
</div>
`;

exports[`TooltipContent OU: none selected 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      None selected
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Period: 2 selected 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      May 2022
    </li>
    <li
      class="item"
    >
      July 2022
    </li>
  </ul>
</div>
`;

exports[`TooltipContent Period: none selected 1`] = `
<div>
  <ul
    class="list"
    data-test="tooltip-content"
  >
    <li
      class="item"
    >
      None selected
    </li>
  </ul>
</div>
`;
